btree/btree.c

```diff
@@ -4466,12 +4492,425 @@ static void m0_btree_ut_multi_stream_kv_oper(void)
+						 &kv_op.bo_op_exec);
+			M0_ASSERT(data.flags == 0);
+
+			keys_put_count++;
+			key_start++;
+		}
+
+		/** GET and ITERATE over the keys which we inserted above. */
+
+		/**  Randomly decide the iteration direction. */
+		random_r(&ti->ti_random_buf, &r);
+
+		key_start = key_iter;
+		if (r%2) {
+			/** Iterate forward. */
+			iter_dir = BOF_NEXT;
+			key[0] = (key_start <<
+				      (sizeof(ti->ti_thread_id) * 8)) +
+				     ti->ti_thread_id;
+			key[0] = m0_byteorder_cpu_to_be64(key[0]);
+			for (i = 1; i < ARRAY_SIZE(key); i++)
+				key[i] = key[0];
+		} else {
+			/** Iterate backward. */
+			iter_dir = BOF_PREV;
+			key[0] = (key_end <<
+				      (sizeof(ti->ti_thread_id) * 8)) +
+				     ti->ti_thread_id;
+			key[0] = m0_byteorder_cpu_to_be64(key[0]);
+			for (i = 1; i < ARRAY_SIZE(key); i++)
+				key[i] = key[0];
+		}
+
+		M0_BTREE_OP_SYNC_WITH_RC(&kv_op.bo_op,
+					 m0_btree_get(tree,
+						      &rec.r_key, &ut_get_cb,
+						      iter_dir, &kv_op),
+					 &kv_op.bo_sm_group,
+					 &kv_op.bo_op_exec);
+		keys_found_count++;
+
+		while (get_data.flags != M0_BSC_KEY_NOT_FOUND) {
+			M0_BTREE_OP_SYNC_WITH_RC(&kv_op.bo_op,
+						 m0_btree_iter(tree,
+							       &rec.r_key,
+							       &ut_get_cb,
+							       iter_dir,
+							       &kv_op),
+						 &kv_op.bo_sm_group,
+						 &kv_op.bo_op_exec);
+			keys_found_count++;
+
+			/** Copy over the gotten key for the next search. */
+			for (i = 1; i < ARRAY_SIZE(key); i++)
+				key[i] = get_key[0];
+		}
+
+		M0_ASSERT(keys_found_count == keys_put_count);
+
+		/**
+		 *  DEL the keys which we had created in this iteration. The
+		 *  direction of traversing the delete keys is randomly
+		 *  selected.
+		 */
+		random_r(&ti->ti_random_buf, &r);
+
+		key_start = key_iter;
+		del_key = (r % 2 == 0) ? key_start : key_end;
+		while (keys_found_count) {
+			key[0] = (del_key << (sizeof(ti->ti_thread_id) * 8)) +
+				 ti->ti_thread_id;
+			key[0] = m0_byteorder_cpu_to_be64(key[0]);
+			for (i = 1; i < ARRAY_SIZE(key); i++)
+				key[i] = key[0];
+
+			M0_BTREE_OP_SYNC_WITH_RC(&kv_op.bo_op,
+						 m0_btree_del(tree, &rec.r_key,
+							      &ut_cb, 0,
+							      &kv_op),
+						 &kv_op.bo_sm_group,
+						 &kv_op.bo_op_exec);
+			del_key = (r % 2 == 0) ? del_key + 1 : del_key - 1;
+			keys_found_count--;
 		}
+
+		key_iter = key_end + 1;
+	}
+
+	/** Free resources. */
+	m0_free(ti->ti_rnd_state_ptr);
+}
+
+
+/**
+ * This function allocates an array pointed by cpuid_ptr and fills it with the
+ * CPU ID of the CPUs which are currently online.
+ */
+static void online_cpu_id_get(uint16_t **cpuid_ptr, uint16_t *cpu_count)
+{
+	size_t           cpu_max;
+	uint32_t         cpuid;
+	struct m0_bitmap map_cpu_online  = {};
+	int              rc;
+
+	*cpu_count = 0;
+	cpu_max = m0_processor_nr_max();
+	rc = m0_bitmap_init(&map_cpu_online, cpu_max);
+	if (rc != 0)
+		return;
+
+	m0_processors_online(&map_cpu_online);
+
+	for (cpuid = 0; cpuid < map_cpu_online.b_nr; cpuid++) {
```

Comment:
This would have been nice but it will not work because we cannot assume that all the CPUs will start from offset 0 and end at TOTAL_CPU_COUNT-1. 
This bitmap array could have some bits which will be '0s' where the CPUs will not be ONLINE even if they are available in the system. While this does not concern our R1/R2 hardware but in NUMA setups this can be a possibility if there are CPUs on expansion cards which might have been turned OFF at this point in time for power reasons.